tools/codegen/model.py

```diff
@@ -529,6 +549,106 @@ def is_foreach_op(name: str) -> bool:
         '_foreach_addcdiv_.ScalarList',
         '_foreach_zero_'])
 
+@dataclass(frozen=True)
+class BackendMetadata:
+    # The name of the backend kernel, for a given operator
+    # for in-tree backends. These names come directly from the 'dispatch" field
+    # in native_functions.yaml. The dispatch entry is optional; in that
+    # case, that is equivalent to having written:
+    #
+    #   dispatch:
+    #       CompositeImplicitAutograd: $operator_name
+    kernel: str
+    # Whether or not the operator has a structured kernel implemented, for this particular backend.
+    # For in-tree backends, they all have the same value for structured- this is listed
+    # in native_functions.yaml.
+    # However, external backends like XLA can indendently toggle which ops are structured.
+    structured: bool
+    # Whether or not this op is an in-tree (e.g. CPU/CUDA) or out-of-tree backend (e.g. XLA)
+    external: bool
```

Comment:
If you ever get around to making that global backend metadata, external probably belongs there, not here.

Reply:
...maybe this should be in BackendIndex? (This is not worth holding up this PR for)

Reply:
Yep, definitely should be in backend index